### PR TITLE
Add test cases to test hotplug maximum number of vCPU devices

### PR DESCRIPTION
**Pull request by PaulYuuu**
Add test cases to test hotplug maximum number of vCPU devices and then hotunplug them.

Bug ID: 1786501
Signed-off-by: Yihuang Yu <yihyu@redhat.com>

### CODE REVIEW

**Comment by PaulYuuu** [comment]
@huangyum Request for your review, thanks a lot.

File: qemu/tests/cfg/cpu_device_hotplug_maximum.cfg [hunk 1]
```diff
@@ -0,0 +1,38 @@
+- cpu_device_hotplug_maximum:
+    virt_test_type = qemu
+    type = cpu_device_hotplug_maximum
+    no ovmf
+    no RHEL.6
+    only Linux
```
Comment (huangyum) [inline qemu/tests/cfg/cpu_device_hotplug_maximum.cfg]: We need test windows guest as well,  there is a bug only reproducible with windows guest.

File: qemu/tests/cfg/cpu_device_hotplug_maximum.cfg [hunk 2]
```diff
@@ -0,0 +1,38 @@
...
+            vcpu_cores = 1
+            vcpu_threads = 0
+            ppc64, ppc64le:
+                smp = 8
+                vcpu_threads = 8
+                vcpu_cores = 0
+    variants:
+        - @default:
+        - with_hugepages:
+            hugepage = yes
```
Comment (huangyum) [inline qemu/tests/cfg/cpu_device_hotplug_maximum.cfg]: setup_hugepages = yes

Comment (PaulYuuu) [reply qemu/tests/cfg/cpu_device_hotplug_maximum.cfg]: when `hugepage = yes`, `setup_hugepages` will be `yes`
https://github.com/avocado-framework/avocado-vt/blob/363a95460c11e86b9feafadeb4e15a409f6a83ca/virttest/env_process.py#L1120-L1121

File: qemu/tests/cfg/cpu_device_hotplug_maximum.cfg (hunk 1 above)
Comment (PaulYuuu) [reply qemu/tests/cfg/cpu_device_hotplug_maximum.cfg]: Could you help to provide CPU limit of each windows version, I don't know much about them.

Comment (huangyum) [reply qemu/tests/cfg/cpu_device_hotplug_maximum.cfg]: Seems they don't have a hard limitation on total cpu number, but the maximum socket number is 64.

Comment (PaulYuuu) [reply qemu/tests/cfg/cpu_device_hotplug_maximum.cfg]: I see that the maximum number of CPU for some windows is 128, so I consider limiting the number of windows CPU to 128. Already updated.

**Comment by PaulYuuu** [comment]
@huangyum Updated code after our discussion.
Result on x86:
```
 (01/12) Host_RHEL.m7.u8.qcow2.virtio_scsi.up.virtio_net.Guest.RHEL.7.8.x86_64.io-github-autotest-qemu.cpu_device_hotplug_maximum.max_socket: PASS (1008.27 s)
```
You can pick whether to overcommit host CPUs, see L40-L41.

result on ppc:
```
 (01/12) Host_RHEL.m8.u2.product_rhel.qcow2.virtio_scsi.up.virtio_net.Guest.RHEL.8.2.0.x86_64.io-github-autotest-qemu.cpu_device_hotplug_maximum.max_socket: PASS (495.71 s)
 (02/12) Host_RHEL.m8.u2.product_rhel.qcow2.virtio_scsi.up.virtio_net.Guest.RHEL.8.2.0.x86_64.io-github-autotest-qemu.cpu_device_hotplug_maximum.max_core: PASS (475.71 s)
 (03/12) Host_RHEL.m8.u2.product_rhel.qcow2.virtio_scsi.up.virtio_net.Guest.RHEL.8.2.0.x86_64.io-github-autotest-qemu.cpu_device_hotplug_maximum.max_thread: PASS (490.78 s)
 (04/12) Host_RHEL.m8.u2.product_rhel.qcow2.virtio_scsi.up.virtio_net.Guest.RHEL.8.2.0.x86_64.io-github-autotest-qemu.cpu_device_hotplug_maximum.with_hugepages.max_socket: PASS (501.45 s)
 (05/12) Host_RHEL.m8.u2.product_rhel.qcow2.virtio_scsi.up.virtio_net.Guest.RHEL.8.2.0.x86_64.io-github-autotest-qemu.cpu_device_hotplug_maximum.with_hugepages.max_core: PASS (503.01 s)
 (06/12) Host_RHEL.m8.u2.product_rhel.qcow2.virtio_scsi.up.virtio_net.Guest.RHEL.8.2.0.x86_64.io-github-autotest-qemu.cpu_device_hotplug_maximum.with_hugepages.max_thread: PASS (504.08 s)
 (07/12) Host_RHEL.m8.u2.product_rhel.qcow2.virtio_scsi.up.virtio_net.Guest.Win2019.x86_64.io-github-autotest-qemu.cpu_device_hotplug_maximum.max_socket: PASS (301.02 s)
 (08/12) Host_RHEL.m8.u2.product_rhel.qcow2.virtio_scsi.up.virtio_net.Guest.Win2019.x86_64.io-github-autotest-qemu.cpu_device_hotplug_maximum.max_core: PASS (312.97 s)
 (09/12) Host_RHEL.m8.u2.product_rhel.qcow2.virtio_scsi.up.virtio_net.Guest.Win2019.x86_64.io-github-autotest-qemu.cpu_device_hotplug_maximum.max_thread: PASS (295.70 s)
 (10/12) Host_RHEL.m8.u2.product_rhel.qcow2.virtio_scsi.up.virtio_net.Guest.Win2019.x86_64.io-github-autotest-qemu.cpu_device_hotplug_maximum.with_hugepages.max_socket: PASS (304.39 s)
 (11/12) Host_RHEL.m8.u2.product_rhel.qcow2.virtio_scsi.up.virtio_net.Guest.Win2019.x86_64.io-github-autotest-qemu.cpu_device_hotplug_maximum.with_hugepages.max_core: PASS (303.64 s)
 (12/12) Host_RHEL.m8.u2.product_rhel.qcow2.virtio_scsi.up.virtio_net.Guest.Win2019.x86_64.io-github-autotest-qemu.cpu_device_hotplug_maximum.with_hugepages.max_thread: PASS (308.70 s)
```

**Comment by PaulYuuu** [comment]
@huangyum avocado-vt force `vcpu_threads = 1` on AMD, so I also cancelled max_thread on AMD.
```
'1-Host_RHEL.m8.u2.product_av.qcow2.virtio_scsi.up.virtio_net.Guest.RHEL.8.2.0.x86_64.io-github-autotest-qemu.cpu_device_hotplug_maximum.max_socket' Finshed 2020-02-19 05:15:42 2020-02-19 05:22:56 PASS 434.157882
'2-Host_RHEL.m8.u2.product_av.qcow2.virtio_scsi.up.virtio_net.Guest.RHEL.8.2.0.x86_64.io-github-autotest-qemu.cpu_device_hotplug_maximum.max_core' Finshed 2020-02-19 05:22:56 2020-02-19 05:30:06 PASS 430.223171
'3-Host_RHEL.m8.u2.product_av.qcow2.virtio_scsi.up.virtio_net.Guest.RHEL.8.2.0.x86_64.io-github-autotest-qemu.cpu_device_hotplug_maximum.max_thread' Finshed 2020-02-19 05:30:06 2020-02-19 05:30:08 CANCEL 2.181492
'4-Host_RHEL.m8.u2.product_av.qcow2.virtio_scsi.up.virtio_net.Guest.RHEL.8.2.0.x86_64.io-github-autotest-qemu.cpu_device_hotplug_maximum.with_hugepages.max_socket' Finshed 2020-02-19 05:30:08 2020-02-19 05:37:19 PASS 430.566219
'5-Host_RHEL.m8.u2.product_av.qcow2.virtio_scsi.up.virtio_net.Guest.RHEL.8.2.0.x86_64.io-github-autotest-qemu.cpu_device_hotplug_maximum.with_hugepages.max_core' Finshed 2020-02-19 05:37:19 2020-02-19 05:44:36 PASS 437.20158
'6-Host_RHEL.m8.u2.product_av.qcow2.virtio_scsi.up.virtio_net.Guest.RHEL.8.2.0.x86_64.io-github-autotest-qemu.cpu_device_hotplug_maximum.with_hugepages.max_thread' Finshed 2020-02-19 05:44:36 2020-02-19 05:44:43 CANCEL 7.150079
```

File: qemu/tests/cpu_device_hotplug_maximum.py [hunk 3]
```diff
@@ -0,0 +1,100 @@
...
+              host_kernel_ver not in VersionInterval("[4, )")):
+            supported_maxcpus = 256
+    vcpu_devices = vcpu_devices or ["vcpu%d" % vcpu for vcpu in
+                                    range(1, supported_maxcpus)]
+    params["vcpu_maxcpus"] = str(supported_maxcpus)
+    params["vcpu_devices"] = " ".join(vcpu_devices)
+    params["start_vm"] = "yes"
+
+    vm = env.get_vm(params["main_vm"])
+    vm.create(params=params)
```
Comment (xutian) [inline qemu/tests/cpu_device_hotplug_maximum.py]: verify_alive in here, it will  save  login time if VM is hang.

File: qemu/tests/cpu_device_hotplug_maximum.py [hunk 4]
```diff
@@ -0,0 +1,100 @@
...
+    session = vm.wait_for_login()
```
Comment (xutian) [inline qemu/tests/cpu_device_hotplug_maximum.py]: if there many vcpus guest will spent lots of time to sync, it maybe take long time, so suggest to set timeout for wait_for_login function.

File: qemu/tests/cpu_device_hotplug_maximum.py [hunk 5]
```diff
@@ -0,0 +1,100 @@
...
+        logging.error(not_equal_text, vm.get_cpu_count(), supported_maxcpus)
+        test.fail(mismatch_text)
+    logging.info("CPU quantity is as expected: %s", supported_maxcpus)
+
+    error_context.context("Check CPU topology of guest", logging.info)
+    if not cpu_utils.check_guest_cpu_topology(session, os_type, cpuinfo):
+        test.fail("CPU topology of guest is not as expected.")
+    session = vm.reboot(session)
+    if not cpu_utils.check_guest_cpu_topology(session, os_type, cpuinfo):
+        test.fail("CPU topology of guest is not as expected.")
```
Comment (xutian) [inline qemu/tests/cpu_device_hotplug_maximum.py]: update message to 'CPU topology of guest is not as expected after reboot' , duplicate information can be an obstacle in debugging.

File: qemu/tests/cpu_device_hotplug_maximum.py (hunk 4 above)
Comment (PaulYuuu) [reply qemu/tests/cpu_device_hotplug_maximum.py]: Not needed here, guest only has 1 CPU before hotplug, but may take a long time during reboot phase if host's CPU is heavily overcommit. But it is not recommended for users, I would consider adding reboot timeout.

File: qemu/tests/cpu_device_hotplug_maximum.py (hunk 3 above)
Comment (PaulYuuu) [reply qemu/tests/cpu_device_hotplug_maximum.py]: Yes it should be added.

File: qemu/tests/cpu_device_hotplug_maximum.py (hunk 5 above)
Comment (PaulYuuu) [reply qemu/tests/cpu_device_hotplug_maximum.py]: will update

**Comment by xutian** [comment]
ACK
